File: backend/app/schemas/coil_price.py
"""Coil price request/response DTOs.

All list/sort/filter field names must stay in sync with
``utils/coil_price/query.py`` and the ``CoilPrice`` model.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from .common import PageQuery

# ---------------------------------------------------------------------------
# Sort key whitelist (Literal → unknown values rejected at parse time)
# ---------------------------------------------------------------------------

CoilPriceSortBy = Literal["quantity", "price", "active", "created_at", "updated_at"]


# ---------------------------------------------------------------------------
# Query DTOs
# ---------------------------------------------------------------------------


class CoilPriceListQuery(PageQuery):
    """Query params for ``GET /admin/coil-prices``.

    Extends ``PageQuery`` (page, limit, sortOrder).  Sort keys are a Pydantic
    ``Literal`` whitelist so unknown values are rejected at parse time
    (backend-api-standards).
    """

    sortBy: CoilPriceSortBy = "quantity"
    # sortOrder, page, limit, skip inherited from PageQuery.
    # None = no filter applied; FastAPI coerces ?active=true/false to bool.
    active: bool | None = None


# ---------------------------------------------------------------------------
# Mutation DTOs
# ---------------------------------------------------------------------------


class CoilPriceCreate(BaseModel):
    """Body for ``POST /admin/coil-prices``."""

    quantity: float = Field(gt=0)
    price: float = Field(ge=0)
    active: bool = True


class CoilPriceUpdate(BaseModel):
    """Body for ``PATCH /admin/coil-prices/{coil_price_id}``.

    All fields are optional so partial updates are valid.
    """

    quantity: float | None = Field(default=None, gt=0)
    price: float | None = Field(default=None, ge=0)
    active: bool | None = None


# ---------------------------------------------------------------------------
# Response DTOs
# ---------------------------------------------------------------------------


class CoilPricePublic(BaseModel):
    """Client-safe projection of a CoilPrice document.

    ``id`` is the MongoDB ObjectId as a plain string so the frontend never
    touches ObjectId logic (``api-contract-standards``, ``owasp-security``).
    """

    id: str
    quantity: float
    price: float
    active: bool
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/admin/coil-price.ts
/**
 * Admin coil price API contracts — aligned with the backend `coil_price` domain.
 * Backs the admin "Coil Config" page's "Per Coil Price" section.
 */

import type { PageQuery } from "@/types/api/envelope"

/** Whitelisted sort keys for `GET /admin/coil-prices` (must match backend `CoilPriceSortBy` Literal). */
export type CoilPriceSortBy = "quantity" | "price" | "active" | "created_at" | "updated_at"

/**
 * Client-safe coil price projection returned by all admin coil-price endpoints.
 * Mirrors backend `CoilPricePublic` schema. Timestamps are snake_case (backend
 * serializes directly, matching the Region domain).
 */
export interface CoilPrice {
  id: string
  quantity: number
  price: number
  active: boolean
  /** ISO-8601 UTC timestamp. */
  created_at: string
  /** ISO-8601 UTC timestamp. */
  updated_at: string
}

/**
 * Query params for `GET /admin/coil-prices`.
 * All filtering/sorting/pagination is server-driven — never apply client-side.
 */
export interface CoilPriceListQuery extends PageQuery {
  sortBy?: CoilPriceSortBy
  /** "all" = UI sentinel, stripped before API request. */
  active?: boolean | "all"
}

/** Request body for `POST /admin/coil-prices`. */
export interface CreateCoilPriceInput {
  quantity: number
  price: number
  active: boolean
}

/**
 * Request body for `PATCH /admin/coil-prices/{id}`.
 * All fields are optional — send only the fields being changed.
 */
export interface UpdateCoilPriceInput {
  quantity?: number
  price?: number
  active?: boolean
}
